Fix #4065: New  editor logout handler

akumar1503 opened the pull request:
- created new logout handler ''EditorLogoutHandler" in editor.py.

- added route in  "main.py".

- modified "create_logout_url()" in "gae_current_user_services.py" to check for exp_editor page and process accordingly.

Testing:
- login and click on create. now logout: should redirect to /library.

- login and click on create. publish the exploration. now logout: should redirect to same page.

akumar1503 committed:
logout from exp_editor working as expected.

akumar1503 committed:
fixed linting

akumar1503 committed:
fixed linting

akumar1503 committed:
fixed linting

seanlip commented on core/controllers/base_test.py:
```diff
@@ -519,7 +519,8 @@ def test_every_method_has_decorator(self):
             if (handler.__name__ == 'LogoutPage' or
-                    handler.__name__ == 'Error404Handler'):
+                    handler.__name__ == 'Error404Handler' or
+                    handler.__name__ == 'EditorLogoutHandler'):
```
Why add it here? It doesn't have the problem that's noted in the comment above...

akumar1503 replied:
i didn't noticed that, i will create a new if statement if needed.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,6 +101,22 @@ def _require_valid_version(version_from_payload, exploration_version):
+            exploration_id = return_url.split('/')[-1]
+            exp_summary = exp_services.get_exploration_summary_by_id(
+                exploration_id)
+            if exp_summary.status == 'private':
```
Use the named constant ACTIVITY_STATUS_PRIVATE instead.

seanlip commented on core/platform/users/gae_current_user_services.py:
```diff
@@ -32,8 +32,14 @@ def create_login_url(slug):
-    logout_url = utils.set_url_query_parameter('/logout', 'return_url', slug)
+    """Creates a logout url.
+    Logout from exploration editor is handled separately."""
+    if 'create' in slug.split('/'):
```
This isn't the right place to do this -- the platform layer shouldn't know about 'create'. It needs to be done somewhere higher up, perhaps in the editor page controller.

akumar1503 replied:
i didn't got a way to change logout href without modifing 'create_logout_url()'.

the other way could be to handle exp_editor logout in LogoutPage(as earlier) and check if it's from exp_editor, but this would require importing exp_services in base.py.

i think this is fine, instead of creating a new handler, what are your thought on this.

seanlip replied:
base.py shouldn't import exp_services directly, or even know about the different types of pages. I suppose one way around this is to create a get_logout_url() method in the base controller class and override it in the editor page controller class.

seanlip reviewed (changes requested):
Good start! PTAL at comments. This PR also needs backend tests.

akumar1503 commented:
i will add backend test once it's finalised :)

akumar1503 committed:
handled exp_editor_logout in base.py, removed from editor.py

seanlip commented:
> i will add backend test once it's finalised :)

Nope -- I'm not going to review this again until it has backend tests and the behaviour is verified. Good tests would test the interface, not the implementation, and shouldn't have to change even if the implementation needs subsequent updates. Rather, the backend test would ensure that the reviewer is looking at a PR that's fully working.

akumar1503 committed:
added backend test for logout from exp_editor

akumar1503 committed:
created exp_editor_logout_handler in editor.py and get_logout_url and modified tests

akumar1503 committed:
fixed linting

akumar1503 committed:
fixed linting

akumar1503 committed:
remove import added during dev

akumar1503 commented:
@seanlip i have added backend tests and refactored some codes from my previous commits.
current version looks like this:
- added new logout handler '/exploration_editor_logout' in main.
- added 'get_logout_url()' in base.py
- override 'get_logout_url()' in EditorHandler in editor.py
- added EditorLogoutHandler in editor.py
- added backend test '' in editor_test.py
    1. checks logout from unpublished exp. should redirect to '/library'.
    2. checks logout from published exp. should redirect to same page.

pls have a look/review.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -522,6 +522,11 @@ def test_every_method_has_decorator(self):
                     handler.__name__ == 'Error404Handler'):
                 continue
 
+            # Following handler is present in editor.py where acl_decorators
```
Why can't editor.py import acl_decorators?

akumar1503 replied:
i guess 'EditorLogoutHandler' in editor.py inherits from 'base.LogoutPage' and doesn't have get, post... handlers. so it was failing tests so it may not be required to apply acl_decorators for 'EditorLogoutHandler' also it does not need to inherit from base.BaseHandler.

akumar1503 replied:
yes, acl_decorators should be applied. i have applied acl_decorators now, to check if it's valid exp_id.

seanlip commented on core/controllers/base.py:
```diff
@@ -275,6 +275,10 @@ def render_json(self, values):
         json_output = json.dumps(values, cls=utils.JSONEncoderForHTML)
         self.response.write('%s%s' % (feconf.XSSI_PREFIX, json_output))
 
+    def get_logout_url(self, redirect_url_on_logout):
```
Declare as private: _get_logout_url

akumar1503 replied:
done.

seanlip commented on core/controllers/base.py:
```diff
@@ -275,6 +275,10 @@ def render_json(self, values):
         self.response.write('%s%s' % (feconf.XSSI_PREFIX, json_output))
 
+    def get_logout_url(self, redirect_url_on_logout):
+        """Returns logout url."""
```
Docstring should include Args and Returns sections; follow the format of existing docstrings in files in this codebase.

akumar1503 replied:
done.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,30 @@ def _require_valid_version(version_from_payload, exploration_version):
-    pass
+
+    def get_logout_url(self, redirect_url_on_logout):
+        """Returns logout url for exploration editor."""
```
Should mention in the docstring something about this overriding the superclass method.

Also add Args and Returns sections.

akumar1503 replied:
done.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,30 @@ def _require_valid_version(version_from_payload, exploration_version):
+        return_url = str(self.request.get('return_url'))
+        if return_url:
+            exploration_id = return_url.split('/')[-1]
+            exp_summary = exp_services.get_exploration_summary_by_id(
```
Need to handle the case where the exploration_id is invalid (users can hypothetically enter anything they like in the URL bar) and add a backend test for it.

akumar1503 replied:
yes, will do it.

akumar1503 replied:
i have handled that using '@acl_decorators.can_play_exploration'.

seanlip commented on core/controllers/editor_test.py:
```diff
@@ -412,6 +412,52 @@ def _submit_answer(
+
+    def test_logout_from_unpublished_exploration_editor(self):
+        """Logout from unpublished exploration should redirect
+        to library page."""
```
Move """ to next line, ditto below.

akumar1503 replied:
done.

seanlip commented on core/controllers/editor_test.py:
```diff
@@ -412,6 +412,52 @@ def _submit_answer(
+            unpublished_exp_id)
+        exp_services.save_new_exploration(self.owner_id, exploration)
+
+        current_page = '%s/%s' % (feconf.EDITOR_URL_PREFIX, unpublished_exp_id)
```
current_page_url

Ditto below.

akumar1503 replied:
done

seanlip reviewed (changes requested):
Much better, thanks! This is close to being done, just a few more comments.

akumar1503 committed:
added invalid logout url check and tests, modified editor logout handler.

akumar1503 committed:
modified comments.

akumar1503 commented:
@seanlip i have added the invalid logout url test.
i have applied acl_decorators to check if exp_id is valid and made other requested changes.

seanlip commented on core/controllers/base.py:
```diff
@@ -275,6 +275,18 @@ def render_json(self, values):
+        by LogoutPage handler.
+
+        Args:
+            redirect_url_on_logout: string. URL to redirect to on logout.
```
str, not string, here and below. Use whatever is returned when typing `type(redirect_url_on_logout)` in Python.

akumar1503 replied:
done.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,53 @@ def _require_valid_version(version_from_payload, exploration_version):
+    """Handles logout from editor page."""
+
+    @acl_decorators.can_play_exploration
+    def is_valid_exploration_id(self, exp_id):
```
Decorators are only meant for get(), put(), etc. Drop this method altogether.

akumar1503 replied:
ok. removed it

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,53 @@ def _require_valid_version(version_from_payload, exploration_version):
+        return_url = str(self.request.get('return_url'))
+        if return_url:
+            exploration_id = return_url.split('/')[-1]
+            self.is_valid_exploration_id(exploration_id)
```
Drop this. You should check it in the call below perhaps.

akumar1503 replied:
now i have used 'rights_manager.get_exploration_rights' to check if exp_id is valid.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,53 @@ def _require_valid_version(version_from_payload, exploration_version):
+            exp_summary = exp_services.get_exploration_summary_by_id(
+                exploration_id)
+
+            if exp_summary.status == feconf.ACTIVITY_STATUS_PRIVATE:
```
Minimize the amount of code that's being branched on.
```
url_to_redirect_to = (
    ... if exp_summary.is_private() else ...)
self.redirect(url_to_redirect_to)
```

akumar1503 replied:
i have cleaned to code.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,53 @@ def _require_valid_version(version_from_payload, exploration_version):
+        """Checks if exploration is published and redirects accordingly."""
+
+        return_url = str(self.request.get('return_url'))
+        if return_url:
```
Slightly better to define url_to_redirect_to here (see below), do the if clause, and then do the self.redirect() at the very end (outside all the if/else). Then you don't need the else on line 129. Makes the code easier to understand.

akumar1503 replied:
done.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,53 @@ def _require_valid_version(version_from_payload, exploration_version):
+        EditorLogoutHandler.
+
+        Args:
+            redirect_url_on_logout: string. URL to redirect to on logout.
```
str not string; see above.

akumar1503 replied:
done.

seanlip commented on core/controllers/editor_test.py:
```diff
@@ -412,6 +412,75 @@ def _submit_answer(
+
+        self.login(self.OWNER_EMAIL)
+        response = self.testapp.get(invalid_logout_url, expect_errors=True)
+        self.assertEqual(response.status_int, 404)
```
Why does this result in 404 and not a redirect to the library page?

Also, there are two different logout cases: an empty return_url, and a nonexistent exp id. Test both cases separately.

akumar1503 replied:
i have changed it to redirect to library.

akumar1503 committed:
cleaned code in EditorLogoutHandler, added test for empty redirect url

akumar1503 committed:
fixed linting

akumar1503 commented:
@seanlip i have made the requested changes. code looks more cleaner now.

akumar1503 commented:
sorry i mistakenly clicked close and comment.

seanlip commented on core/controllers/base.py:
```diff
@@ -275,6 +275,18 @@ def render_json(self, values):
+            redirect_url_on_logout: str. URL to redirect to on logout.
+
+        Returns:
+            str. logout url.
```
str. The logout URL.

(Start all descriptions with capital letters and have them form full sentences. Be consistent in terms of using URL (as you've done above) vs url.)

akumar1503 replied:
done.

seanlip commented on core/controllers/editor.py:
```diff
@@ -101,9 +101,45 @@ def _require_valid_version(version_from_payload, exploration_version):
+
+        url_to_redirect_to = str(self.request.get('return_url'))
+        if url_to_redirect_to:
+            exploration_id = url_to_redirect_to.split('/')[-1]
```
Sorry -- I just realized that this would be cleaner if we matched it against a regex representing the expected return_url structure. Can we do that? We shouldn't be processing stuff like /sadjklashkjsdhfdshkg/possible_exp_id.

akumar1503 replied:
@seanlip i have changed path to relative style. i'm using 
```
url_to_redirect_to_regex = (
        	r'%s/(?P<explotation_id>[\w-]+)$' % feconf.EDITOR_URL_PREFIX)
```
regex to match to '/create/exp_id'. is there any character missing for exp_id e.g i think "[\w-]+" should match to every exp_id. can any other characters appear.

seanlip reviewed (commented):
Almost there! Two small things left.

akumar1503 committed:
added regex validation

akumar1503 committed:
fixed linting

akumar1503 committed:
fixed linting

akumar1503 commented:
@seanlip i have added the regex test and other requested changes.

seanlip commented on core/controllers/editor.py:
```diff
@@ -16,6 +16,7 @@
 
 """Controllers for the editor view."""
 
+import re
```
Alphabetize imports.

seanlip commented on core/controllers/editor.py:
```diff
@@ -109,8 +110,12 @@ def get(self):
-        if url_to_redirect_to:
-            exploration_id = url_to_redirect_to.split('/')[-1]
+        url_to_redirect_to_regex = (
+            r'%s/(?P<explotation_id>[\w-]+)$' % feconf.EDITOR_URL_PREFIX)
```
Exploration IDs are strings containing a-zA-Z0-9_-; so please make this check more strict. Also exploration is misspelled here.

Also add a backend test for a bad match.

akumar1503 replied:
[\w] should take care of a-zA-Z0-9_ and i have added -. can you pls suggest other good ways.
for the rest, i will do it soon.

seanlip replied:
Ah, sorry. I misread. What you have is fine.

akumar1503 committed:
added tests for invalid urls

akumar1503 committed:
added invalid logout regex tests

seanlip reviewed (approved):
LGTM. Thanks @GanitGenius!

akumar1503 commented:
@seanlip thank you, was nice working on the issue :)
